src/lib.rs

```diff
@@ -69,39 +89,163 @@ impl<T: AsRef<[u8]>> Mergle<T> {
         let node = MergleInternalNode {
             hash: hash,
             left: Rc::clone(&self.root),
-            right: Rc::clone(&other.root)
+            right: Rc::clone(&other.root),
         };
         Mergle {
             root: Rc::new(MergleNode::Internal(node)),
-            table: Rc::clone(&self.table)
+            table: Rc::clone(&self.table),
+        }
+    }
+
+    pub fn copy(&self) -> Self {
+        Mergle {
+            root: Rc::clone(&self.root),
+            table: Rc::clone(&self.table),
+        }
+    }
+
+    fn values(&self) -> Vec<&T> {
+        self.root.values()
+    }
+
+    fn from_node(node: Rc<MergleNode<T>>, table: MemoizationTableRef<T>) -> Mergle<T> {
+        Mergle {
+            root: node,
+            table: table,
+        }
+    }
+
+    fn leaf_cmp(a: &MergleLeaf<T>, b: &MergleLeaf<T>) -> PrefixResult<T> {
+        match a.content.cmp(&b.content) {
+            Ordering::Less => PrefixResult::LessThan,
+            Ordering::Equal => PrefixResult::Equal,
+            Ordering::Greater => PrefixResult::GreaterThan,
+        }
+    }
+
+    pub fn prefix_cmp(&self, other: &Self) -> PrefixResult<T> {
+        let self_hash = self.root.bromberg_hash();
+        let other_hash = other.root.bromberg_hash();
+        {
+            // isolate immutable borrow
+            let table: &MemoizationTable<T> = &self.table.borrow();
+            if let Some(result) = table.get(&(self_hash, other_hash)) {
+                return (*result).clone();
+            } else if let Some(result) = table.get(&(other_hash, self_hash)) {
+                return result.inverse();
+            }
+        }
+
+        let result = match self.root.as_ref() {
```

Comment:
A trick from Haskell is matching the pair (self.root.as_ref(), other.root.as_ref()); not sure whether that would make this nicer or less nice.

Reply:
Yeah, I think I like that a bit better